File: weather-demo-app-apple/Weather/Data/PreferencesRepository.swift
//
//  UserDefaultsPreferencesStore.swift
//  Weather
//
//  Created by Onur on 4/11/26.
//

import Foundation

final class PreferencesRepository: PreferencesDataSource {
    private let defaults: UserDefaults

    private enum Keys {
        static let lastCity = "last_city"
        static let welcomeSeen = "welcome_seen"
    }

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func saveLastSearchedCity(_ city: String) {
        defaults.set(city, forKey: Keys.lastCity)
    }

    func getLastSearchedCity() -> String? {
        defaults.string(forKey: Keys.lastCity)
    }

    func setWelcomeSeen(_ seen: Bool) {
        defaults.set(seen, forKey: Keys.welcomeSeen)
    }

    func hasSeenWelcome() -> Bool {
        defaults.bool(forKey: Keys.welcomeSeen)
    }
}

File: weather-demo-app-android/app/src/main/java/com/demo/weather/data/PreferencesRepository.kt
package com.demo.weather.data

import android.content.SharedPreferences
import androidx.core.content.edit


class PreferencesRepository(
    private val preferences: SharedPreferences
) : PreferencesDataSource {

    private object Keys {
        const val LAST_CITY = "last_city"
        const val WELCOME_SEEN = "welcome_seen"
    }

    override fun saveLastSearchedCity(city: String) {
        preferences.edit { putString(Keys.LAST_CITY, city) }
    }

    override fun getLastSearchedCity(): String? {
        return preferences.getString(Keys.LAST_CITY, null)
    }

    override fun setWelcomeSeen(seen: Boolean) {
        preferences.edit { putBoolean(Keys.WELCOME_SEEN, seen) }
    }

    override fun hasSeenWelcome(): Boolean {
        return preferences.getBoolean(Keys.WELCOME_SEEN, false)
    }
}